File: bilmarknad-mcp/src/bilmarknad_mcp/urls.py
from __future__ import annotations

import re
from urllib.parse import urlparse

_BLOCKET_ITEM = re.compile(r"/mobility/item/(?P<id>\d+)", re.I)
_WAYKE_VEHICLE = re.compile(r"/(?:vehicle|fordon|bilar)/(?P<id>[^/?#]+)", re.I)
_KVD_OBJECT = re.compile(r"/(?:objekt|vehicle|bil|auktion)/(?P<id>\d+)", re.I)
_TRADERA_ITEM = re.compile(r"/item/(?P<id>\d+)", re.I)
_RIDDERMARK_CAR = re.compile(r"/kopa-bil/(?P<id>[^/?#]+/[^/?#]+)", re.I)
_CARLA_CAR = re.compile(r"/bil/(?P<id>[^/?#]+)", re.I)


def parse_listing_url(url: str) -> tuple[str, str] | None:
    """Detect marketplace source and listing id from a public listing URL."""
    raw = (url or "").strip()
    if not raw:
        return None
    parsed = urlparse(raw)
    host = (parsed.netloc or "").lower().removeprefix("www.")
    path = parsed.path or ""

    if host.endswith("blocket.se"):
        match = _BLOCKET_ITEM.search(path)
        if match:
            return ("blocket", match.group("id"))

    if host.endswith("wayke.se"):
        match = _WAYKE_VEHICLE.search(path)
        if match:
            return ("wayke", match.group("id"))

    if host.endswith("kvd.se"):
        match = _KVD_OBJECT.search(path)
        if match:
            return ("kvd", match.group("id"))

    if host.endswith("tradera.com") or host.endswith("tradera.se"):
        match = _TRADERA_ITEM.search(path)
        if match:
            return ("tradera", match.group("id"))

    if host.endswith("riddermarkbil.se"):
        match = _RIDDERMARK_CAR.search(path)
        if match:
            return ("riddermark", match.group("id").strip("/").lower())

    if host.endswith("carla.se"):
        match = _CARLA_CAR.search(path)
        if match:
            return ("carla", match.group("id").strip("/"))

    return None

File: internal/bilmarknad/urls/urls.go
package urls

import (
	"net/url"
	"regexp"
	"strings"
)

var (
	blocketItem     = regexp.MustCompile(`(?i)/mobility/item/(?P<id>\d+)`)
	waykeVehicle    = regexp.MustCompile(`(?i)/(?:vehicle|fordon|bilar)/(?P<id>[^/?#]+)`)
	kvdObject       = regexp.MustCompile(`(?i)/(?:objekt|vehicle|bil|auktion)/(?P<id>\d+)`)
	traderaItem     = regexp.MustCompile(`(?i)/item/(?P<id>\d+)`)
	riddermarkCar   = regexp.MustCompile(`(?i)/kopa-bil/(?P<make>[^/]+)/(?P<id>[^/?#]+)`)
	carlaCar        = regexp.MustCompile(`(?i)/bil/(?P<id>[^/?#]+)`)
)

// ParseListingURL extracts source and listing id from a public URL.
func ParseListingURL(rawURL string) (source, id string, ok bool) {
	raw := strings.TrimSpace(rawURL)
	if raw == "" {
		return "", "", false
	}
	parsed, err := url.Parse(raw)
	if err != nil {
		return "", "", false
	}
	host := strings.TrimPrefix(strings.ToLower(parsed.Host), "www.")
	path := parsed.Path

	if strings.HasSuffix(host, "blocket.se") {
		if m := blocketItem.FindStringSubmatch(path); len(m) >= 2 {
			return "blocket", m[1], true
		}
	}
	if strings.HasSuffix(host, "wayke.se") {
		if m := waykeVehicle.FindStringSubmatch(path); len(m) >= 2 {
			return "wayke", m[1], true
		}
	}
	if strings.HasSuffix(host, "kvd.se") {
		if m := kvdObject.FindStringSubmatch(path); len(m) >= 2 {
			return "kvd", m[1], true
		}
	}
	if strings.HasSuffix(host, "tradera.com") || strings.HasSuffix(host, "tradera.se") {
		if m := traderaItem.FindStringSubmatch(path); len(m) >= 2 {
			return "tradera", m[1], true
		}
	}
	if strings.HasSuffix(host, "riddermarkbil.se") {
		if m := riddermarkCar.FindStringSubmatch(path); len(m) >= 3 {
			return "riddermark", strings.ToLower(m[1]) + "/" + strings.ToLower(m[2]), true
		}
	}
	if strings.HasSuffix(host, "carla.se") {
		if m := carlaCar.FindStringSubmatch(path); len(m) >= 2 {
			return "carla", m[1], true
		}
	}
	return "", "", false
}
